### PR TITLE
run script Faster

### PR DESCRIPTION
**Pull request by JoesCat**
A lot of code modified with the intent of running faster from start to found (or exit).
Labels were added near the bottom, so that many tests can GOTO direct locations instead of doing further tests that aren't necessary with the existing code.
These locations (below) can be further customized if needed/wanted.

Also added code for issue #259 (needs testing).
Also added code for issue #264 (needs testing).
Also added code for Moto G3.
Also did some edits to Samsung based on some further trial checks.

### CODE REVIEW

**Comment by JoesCat** [comment]
Due to the large number of edits, this is probably worth reviewing, making some further improvements, and maybe waiting a bit before merging.
Some code didn't really know what to do with (yet), therefore left those more or less untouched for now.

**Comment by M0Rf30** [comment]
Hi @JoesCat and thanks for this effort. Can you please split new devices commits onto another PR? Would be nice to have only refactoring commits. Can you also reduce the number of commit? I think a "commit for vendor" is too much. 
Something like:

`refactor: improve test evaluation and general speed improvement`

I'm using conventional commit for commit message in order to ease the changelog generation

File: 51-android.rules [hunk 1]
```diff
@@ -24,87 +24,87 @@ LABEL="android_usb_rules_begin"
 # Acer
 ATTR{idVendor}!="0502", GOTO="not_Acer"
 #   Iconia Tab A1-830
-ATTR{idProduct}=="3604", ENV{adb_adbfast}="yes"
+ATTR{idProduct}=="3604", GOTO="go_adbfast"
```
Comment (M0Rf30) [inline 51-android.rules:27]: labels without go_ prefix would be better imho

**Review by M0Rf30 — CHANGES_REQUESTED** [review]

**Comment by JoesCat** [comment]
It is a big edit - but at least you get an overall idea where it would eventually be headed.

I guess the first part would be to create the labels, like in the first commit (above).
Would it make sense to create separate PRs for all the various vendor groups, or attempt to do a large patch trying to capture as many GOTO in one shot?

In the meantime, if #259 and #264 can test/add their feedback, that will go well - to verify this is going in the right direction.

File: 51-android.rules (hunk 1 above)
Comment (JoesCat) [reply 51-android.rules:27]: It makes it more compact - I agree too.

**Comment by M0Rf30** [comment]
I don't get what you want to say. Big or small doesn't make sense to have commits for every single vendor if it's a refactor.
In this way you will have only one commit related to the refactor (that collect commits from [10ff832](https://github.com/M0Rf30/android-udev-rules/pull/265/commits/10ff83215b154797d511dd8498878cdfd8bf7f5a) to https://github.com/M0Rf30/android-udev-rules/pull/265/commits/d8791d12fd7b12fd0eda9a886f5d01760afcb4e5
and other commits (also in this PR) for new devices (https://github.com/M0Rf30/android-udev-rules/pull/265/commits/11589573b4e0bf09606f981410c3b6afa34315c1)

about #259 and #264 edits: Would be nice to split them entirely it they need testing.

In this way we will test refactoring and issues resolutions separately and avoiding too much entropy

**Comment by JoesCat** [comment]
Started with the one-to-one conversion (just the GOTOs), seems it also answers the "user" issue I was thinking about, but have to leave some of the ENV{} tests as-is for now.
Leaving the vendor related commits for after this first patch, issue #266.

**Comment by mskiptr** [comment]
I guess this will have to evolve into a new patchset on top of [JoesCat/android-udev-rules:goto](https://github.com/JoesCat/android-udev-rules/tree/goto)? Or rather as additional commits that start using all these new labels for existing devices as part of that PR and a new PR that introduces new entries later, if I understand correctly what @M0Rf30 suggested:

> Can you please split new devices commits onto another PR? Would be nice to have only refactoring commits.


Overall, I really like the direction this is going in. It makes it much clearer how new devices (and all their various modes) should be added. After #266 lands I think I'll have no problem with submitting a patch for #264.
And just a heads-up: Moto Z3 is AFAIK a Verizon-only phone with specs different from Moto Z3 Play (it even uses a different SoC). I have no idea if it uses the same USB IDs or not.

**Comment by JoesCat** [comment]
Yes, sometimes it's a bit of effort to make a second try at a problem, and sometimes the second try turns out possibly better (or worse). If #266 looks good and lands, it would be great if you submit a patch. It's a big patch, so give it some time to be digested as good/bad/modified first. Throwing patches in the middle of the changes will end up in a tangled conflict of merges.

**Comment by mskiptr** [comment]
> Throwing patches in the middle of the changes will end up in a tangled conflict of merges.

Indeed! And in such cases rebasing isn't straightforward either.

**Comment by M0Rf30** [comment]
Closing this, replaced by https://github.com/M0Rf30/android-udev-rules/pull/266
Again:
please use conventional commit messages

**Pull request closed without merging**